src/map.cpp

```diff
@@ -529,62 +529,138 @@ bool Map::canThrowObjectTo(const Position& fromPos, const Position& toPos, bool
+bool Map::getSteepLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
 
-		if (start.y != destination.y && (start.x == destination.x || move_hor > move_ver || move_hor > move_cross)) {
-			start.y += my;
+	for (uint16_t y = x0 + 1; y < x1; ++y) {
+		if (!isTileClear(std::floor(xy), y, z)) {
+			return false;
 		}
+		xy += grad;
+	}
 
-		if (start.x != destination.x && (start.y == destination.y || move_ver > move_hor || move_ver > move_cross)) {
-			start.x += mx;
-		}
+	return true;
+}
+
+bool Map::getSlightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	float dx = x1 - x0;
+	float dy = y1 - y0;
+	float grad = (dx == 0) ? 1 : dy / dx;
+	float xy = y0 + grad;
 
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->hasProperty(CONST_PROP_BLOCKPROJECTILE)) {
+	for (uint16_t x = x0 + 1; x < x1; ++x) {
+		if (!isTileClear(x, std::floor(xy + 0.1), z)) {
 			return false;
 		}
+		xy += grad;
 	}
 
-	// now we need to perform a jump between floors to see if everything is clear (literally)
-	while (start.z != destination.z) {
-		const Tile* tile = getTile(start.x, start.y, start.z);
-		if (tile && tile->getThingCount() > 0) {
-			return false;
+	return true;
+}
+
+bool Map::checkSightLine(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1, uint8_t z) const
+{
+	if (x0 == x1 && y0 == y1) {
+		return true;
+	}
+
+	if (std::abs(y1 - y0) > std::abs(x1 - x0)) {
+		if (y1 > y0) {
+			return getSteepLine(y0, x0, y1, x1, z);
 		}
+		return getSteepLine(y1, x1, y0, x0, z);
+	}
 
-		start.z++;
+	if (x0 > x1) {
+		return getSlightLine(x1, y1, x0, y0, z);
 	}
 
-	return true;
+	return getSlightLine(x0, y0, x1, y1, z);
 }
 
-bool Map::isSightClear(const Position& fromPos, const Position& toPos, bool floorCheck) const
+bool Map::isSightClear(const Position& fromPos, const Position& toPos, bool floorCheck /*= false*/) const
 {
-	if (floorCheck && fromPos.z != toPos.z) {
+	uint16_t x0 = fromPos.x;
+	uint16_t y0 = fromPos.y;
+	uint8_t  z0 = fromPos.z;
+
+	uint16_t x1 = toPos.x;
+	uint16_t y1 = toPos.y;
+	uint8_t  z1 = toPos.z;
+
+	//target is on the same floor
+	if (z0 == z1) {
+		//skip checks if toPos is next to us
+		if (Position::getDistanceX(fromPos, toPos) < 2 && Position::getDistanceY(fromPos, toPos) < 2) {
+			return true;
+		}
+
+		//sight is clear or floorCheck is enabled
+		bool sightClear = checkSightLine(x0, y0, x1, y1, z0);
+		if (sightClear || floorCheck) {
+			return sightClear;
+		}
+
+		//no obstacles above floor 0 so we can throw above the obstacle
+		if (z0 == 0) {
+			return true;
+		}
+
+		//check if tiles above us and the target are clear and check for a clear sight between them
+		--z0;
+		return checkSightLine(x0, y0, x1, y1, z0) && isTileClear(x0, y0, z0, true) && isTileClear(x1, y1, z0, true);
```

Comment:
```suggestion
		return isTileClear(x0, y0, z0, true) && isTileClear(x1, y1, z0, true) && checkSightLine(x0, y0, x1, y1, z0);
```

This way you immediately short circuit if the origin or destination are not clear, without checking the entire path.

You can also move the checking of the origin floor to the `checkSteep/SlightLine` functions if you start from `x0` and `y0` instead of `x0 + 1` and `y0 + 1`

Reply:
checking starting and ending spots will bug in the situation when both target and caster are on the same floor so I'll skip this one